### PR TITLE
Ensure that each generic local function gets its own container to cache dynamic call-sites.

### LINKED ISSUE #49599: BUG: Incorect code-gen for local methods along with generics and dynamic
**Version Used**: VS Pro 16.8.2 (targeting .NET Core 3.1 or .NET 5)

**Steps to Reproduce**:
Compile and run:
```C#
public class Program
{
    static void Main()
    {
        Local1<object>();

        static void Local1<T>()
        {
            Local3();

            static int Local2(dynamic n) => (int)n;
        }

        static int Local3() => (int)(dynamic)4;
    }
}
```

**Expected Behavior**:
Program runs successfully.

**Actual Behavior**:
Fails with: System.BadImageFormatException: 'An attempt was made to load a program with an incorrect format.'.

Program runs successfully when at least one of following is true:
- Generic parameter T is removed.
- Definition of Local2 is removed.
- Definition of Local2 is placed next to definition of Local3.
- Definition of Local3 is placed next to definition of Local2.
- Local2 does not take 'dynamic' as param.
- Local3 does not use 'dynamic'.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #49599.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 1]
```diff
@@ -264,7 +264,7 @@ public override BoundNode VisitLambda(BoundLambda node)
 
         public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
         {
-            _sawLocalFunctions = true;
+            int localFunctionOrdinal = _availablelocalFunctionOrdinal++;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:267]: This doesn't need to be atomic, right?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 2]
```diff
@@ -300,19 +300,28 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
...
             {
                 _factory.CurrentFunction = localFunction;
                 if (localFunction.IsDirectlyExcludedFromCodeCoverage)
                 {
                     _instrumenter = RemoveDynamicAnalysisInjectors(oldInstrumenter);
                 }
+
+                if (localFunction.IsGenericMethod)
+                {
+                    // Each generic local function gets its own dynamic factory
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:314]: Consider expanding this to explain why. As is, the comment isn't very helpful.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs [hunk 3]
```diff
@@ -5937,6 +5937,81 @@ static void validate(ModuleSymbol module)
...
+            System.Console.Write(Local2<object>(Local3<object>()));	
+            static string Local2<U>(dynamic n) => n.ToString();
+        }
+
+        static int Local3<S>() => (int)(dynamic)4;
+    }
+}
+";
+            CompileAndVerify(source, targetFramework: TargetFramework.StandardAndCSharp, expectedOutput: "44");
+        }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs:6013]: Consider adding a test case where an inner local function shadows the name of an outer local function.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 2)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:267]: > For my own clarifiction, this doesn't need to be atomic, right?

It is atomic, I believe.

---
In reply to: [537700973](https://github.com/dotnet/roslyn/pull/49802#discussion_r537700973) [](ancestors = 537700973)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs:6013]: > Consider adding a test case where an inner local function shadows the name of an outer local function.

Are you asking just out of completeness? As far as I can tell names or shadowing do not gave any effect on the scenario.

---
In reply to: [537703406](https://github.com/dotnet/roslyn/pull/49802#discussion_r537703406) [](ancestors = 537703406)

Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs:6013]: Just out of completeness.

**Comment by AlekseyTs** [comment]
@333fred I think I addressed feedback.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 4]
```diff
@@ -25,9 +25,9 @@ internal sealed partial class LocalRewriter : BoundTreeRewriterWithStackGuard
         private readonly SyntheticBoundNodeFactory _factory;
         private readonly SynthesizedSubmissionFields _previousSubmissionFields;
         private readonly bool _allowOmissionOfConditionalCalls;
-        private readonly LoweredDynamicOperationFactory _dynamicFactory;
+        private LoweredDynamicOperationFactory _dynamicFactory;
         private bool _sawLambdas;
-        private bool _sawLocalFunctions;
+        private int _availablelocalFunctionOrdinal;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:30]: nit: it seems like this variable should be named `_availableLocalFunctionOrdinal`

**Review by RikkiGibson — APPROVED** [review]

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:30]: > it seems like this variable should be named _availableLocalFunctionOrdinal

I'll take care of this in a follow up PR.

---
In reply to: [538671479](https://github.com/dotnet/roslyn/pull/49802#discussion_r538671479) [](ancestors = 538671479)

**Pull request merged**